src/resolve/resolved-dns-stub.c

```diff
@@ -726,9 +723,15 @@ static int dns_stub_patch_bypass_reply_packet(
                 DNS_PACKET_HEADER(c)->flags = htobe16(be16toh(DNS_PACKET_HEADER(c)->flags) | DNS_PACKET_FLAG_TC);
         }
 
+        /* Patch the cd bit to reflect the state of validation */
+        DNS_PACKET_HEADER(c)->flags = htobe16(
+                        (be16toh(DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_CD) |
+                        (checking_disabled ? DNS_PACKET_FLAG_CD : 0));
+
         /* Ensure we don't pass along an untrusted ad flag for bypass packets */
-        if (!authenticated)
-                DNS_PACKET_HEADER(c)->flags = htobe16(be16toh(DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_AD);
+        DNS_PACKET_HEADER(c)->flags = htobe16(
+                        (be16toh(DNS_PACKET_HEADER(c)->flags) & ~DNS_PACKET_FLAG_AD) |
+                        (!checking_disabled && authenticated ? DNS_PACKET_FLAG_AD : 0));
```

Comment:
ditto

Reply:
On second thought, I think it's fine to have the reply with +cd and +ad in the case the client request is +cd and we happen to return authenticated data from the cache. We won't be doing that because of the SD_RESOLVED_NO_CACHE being set for +cd packets, but still I don't think it's necessary to check both conditions here, so I'll update it to just the `authenticated` value.